sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+    nonceForeignPtr <- BS.mallocByteString @CChar (fromIntegral cryptoSecretboxNonceBytes)
+    Foreign.withForeignPtr nonceForeignPtr $ \noncePtr ->
+      Foreign.copyArray outsideNoncePtr noncePtr (fromIntegral cryptoSecretboxNonceBytes)
+    pure $ Nonce (Foreign.castForeignPtr @CChar @CUChar nonceForeignPtr)
+
+-- | Convert a 'Nonce' to a strict, hexadecimal-encoded 'Text'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+nonceToHexText :: Nonce -> Text
+nonceToHexText = Base16.encodeBase16 . nonceToBinary
+
+-- | Convert a 'Nonce' to a strict, hexadecimal-encoded 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+nonceToHexByteString :: Nonce -> StrictByteString
+nonceToHexByteString = Base16.encodeBase16' . nonceToBinary
+
+-- | Convert a 'Nonce' to a strict binary 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+nonceToBinary :: Nonce -> StrictByteString
+nonceToBinary (Nonce nonceForeignPtr) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr @CUChar @Word8 nonceForeignPtr)
+    (fromIntegral @CSize @Int cryptoSecretboxKeyBytes)
+
+-- | A ciphertext consisting of an encrypted message and an authentication tag.
+--
+-- @since 0.0.1.0
+data Hash = Hash
+  { messageLength :: CULLong
+  , hashForeignPtr :: ForeignPtr CUChar
+  }
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Hash where
+  (Hash messageLength1 hk1) == (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrEq hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ (messageLength1 == messageLength2) && result1
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Hash where
+  compare (Hash messageLength1 hk1) (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrOrd hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ compare messageLength1 messageLength2 <> result1
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display Hash where
+  displayBuilder = Builder.fromText . hashToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show Hash where
+  show = BS.unpackChars . hashToHexByteString
+
+-- | Create a 'Hash' from a 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+hashFromByteString :: StrictByteString -> Hash
+hashFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideHashPtr, outsideHashLength) -> do
+    hashForeignPtr <- BS.mallocByteString @CChar outsideHashLength
+    Foreign.withForeignPtr hashForeignPtr $ \hashPtr ->
+      Foreign.copyArray outsideHashPtr hashPtr outsideHashLength
+    pure $
+      Hash
+        (fromIntegral @Int @CULLong outsideHashLength)
+        (Foreign.castForeignPtr @CChar @CUChar hashForeignPtr)
```

Comment:
How is it encoded? Will _any_ `StrictByteString` do?

Reply:
It's the concatenation of the message + authentication tag, then encrypted. I guess the only size constant here is the MAC.

Reply:
That suggests it could fail.